### DISCUSSION TITLE
Holepunching issues: "failed to negotiate protocol: protocols not supported: [/libp2p/dcutr]" even though it is enabled

### DISCUSSION THREAD

**Discussion by felix314159** [libp2p/go-libp2p#2824] [Q&A]
Hi, I am trying to get holepunching to work. I have made three code examples and a more detailed writeup of my struggles here: https://github.com/felix314159/golibp2p-holepunchingTest

So basically I am running my node with "export GOLOG_LOG_LEVEL=p2p-holepunch=debug" and it shows that other nodes seem to be able to holepunch through to me, but when I run my own code on two computers in different networks I always get "failed to negotiate protocol: protocols not supported: [/libp2p/dcutr]". The main problem for me here is that I am building a gossipsub-based application and it seems that no topic messages can be sent or received by nodes that are not in the same network as me (and I think it is due to the holepunching issues). Can someone look at the node setup in the code I linked and check on your machine if you get the same issues? I do not understand how at the same time it claims dcutr is not supported while it also prints successful holepunches from nodes I do not control.. I have seen multiple holepunch related discussions popping up in the last few weeks and I wonder if they are related.

Thanks for any reply, have a nice day!

**Comment by MarcoPolo** [comment]
Usually if you see `failed to negotiate protocol: protocols not supported: [/libp2p/dcutr]` it's because your node hasn't gotten enough observations from to see if it has a public address (even if it's natd/firewalled). This happens here: https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/holepunch/svc.go#L109. Does this happen even after running your node for a while or only at the beginning?

---

Haven't taken a close look yet at your code, but some qs:
1. Are you sure you have a relay reservation with reliableIPFSserver? It's not clear to me you do.
2. Can you find your node's peer id with ipfs routing findpeer ? This tells you if you are in the dht
3. If you manually use node's A relayed addr from node B to connect, do you end up with a direct connection? Or does holepunching fail?

**Reply by felix314159** [reply]
Hi, thanks for your reply.

0. "Does this happen even after running your node for a while or only at the beginning?" Even when my two nodes have a relayed connection for  30min holepunching still fails with "protocols not supported: [/libp2p/dcutr]"
1. I don't know what 'reservation' means here, so I probably do not have a reserveration. Is there a code example?
2. Yes, this works.
3. If I use a tool like [vole](https://github.com/ipfs-shipyard/vole) with node A's relayed addr it will show how many ms it took (so it is able to ping the other node) but the other node (which vole has targeted with its ping) will print that the holepunching failed.

I also want to add that when I use [libp2p-lookup](https://github.com/mxinden/libp2p-lookup/tree/master) it will print all supported protocols of my node and that includes [/libp2p/dcutr]. So how can it be possible that the node prints "protocols not supported: [/libp2p/dcutr]" when I try to holepunch it. And why are other nodes able to holepunch my node successfully?

Regards
